### PR TITLE
[4.0] Use is_write_window in eos-vm-oc main to determine if it is safe to update code cache

### PR DESCRIPTION
**Pull request by linh2931**
Resolve https://github.com/AntelopeIO/leap/issues/951.

We used to use `is_main_thread` to determine if it was safe to update eos-vm-cc code cache, as a transaction on the main thread did not run in parallel with read-only threads. Now a read-only transaction on the main thread can run in parallel with read-only threads. It is not safe for a read-only transaction on the main thread to modify code cache data structures. The solution is to use app's execution window. When in write window, all transactions are executed sequentially and no read-only threads are executing transactions.

### CODE REVIEW

File: libraries/chain/controller.cpp [hunk 1]
```diff
@@ -2694,6 +2701,16 @@ struct controller_impl {
...
 
+   void set_to_write_window() {
+      app_window = app_window_type::write;
+   }
+   void set_to_read_window() {
+      app_window = app_window_type::read;
+   }
+   bool is_write_window() const {
+      return app_window == app_window_type::write;
+   }
```
Comment (heifner) [inline libraries/chain/controller.cpp:2712]: I think you should only have one of `set_db_read_only_mode` and `set_to_read_window`. Can you just use the db read only state? `context.control.db().is_read_only_mode()` Would need to add the `is_read_only_mode()` accessor to chainbase.

**Review by heifner — CHANGES_REQUESTED** [review]

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -251,6 +257,7 @@ struct controller_impl {
    // which overwrites the same static wasmif, is used for eosvmoc too.
    wasm_interface  wasmif;  // used by main thread and all threads for EOSVMOC
    thread_local static std::unique_ptr<wasm_interface> wasmif_thread_local; // a copy for each read-only thread, used by eos-vm and eos-vm-jit
+   app_window_type app_window;
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/controller.cpp:260]: Thank you so much for checking my tests. I was away from my desk for a while. Will fix it now.

File: libraries/chain/controller.cpp (hunk 1 above)
Comment (linh2931) [reply libraries/chain/controller.cpp:2712]: We need to keep the both. `db_read_only_mode` is for protecting chainbase not to be updated while a read-only transaction is executed, either in read window or write window. It cannot tell whether the app() is in write window (tasks are running sequentially and safe to update global data like eos-vm-oc code cache) or in read window (tasks are running in parallel to each other).

Comment (heifner) [reply libraries/chain/controller.cpp:2712]: That makes sense. I'm not thinking of anyway to make this cleaner at the moment. I do think this is better than a registered function that calls `app().executor().is_write_window()`. 

You can change all the `producer_plugin` calls of `app().executor().is_write_window()` with `chain.is_write_window()`.

**Comment by spoonincode** [comment]
I think this should be pointed to `release/4.0` branch?

**Comment by linh2931** [comment]
Thanks for noticing the bad base branch! Sorry about it. Too many PRs today.

**Review by spoonincode — APPROVED** [review]

File: libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp [hunk 3]
```diff
@@ -136,11 +137,11 @@ const code_descriptor* const code_cache_async::get_descriptor_for_code(const dig
...
    code_cache_index::index<by_hash>::type::iterator it = _cache_index.get<by_hash>().find(boost::make_tuple(code_id, vm_version));
    if(it != _cache_index.get<by_hash>().end()) {
-      if (is_main_thread())
+      if (is_write_window)
          _cache_index.relocate(_cache_index.begin(), _cache_index.project<0>(it));
       return &*it;
    }
-   if(!is_main_thread()) // on read-only thread
+   if(!is_write_window)
       return nullptr;
```
Comment (spoonincode) [inline libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp:145]: one initial concern I had was what would happen for a contract _only_ used for ROtrxs -- would it never get tiered up? @heifner pointed out to me that we do run ROtrxs during the write window as a low priority task, so it seems like a contract being used only for ROtrxs would indeed have an opportunity to be tiered up during that period

**Review by heifner — APPROVED** [review]

File: libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp (hunk 3 above)
Comment (linh2931) [reply libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp:145]: Thanks for digging this out. I had another concern last night when re-reading the code and was going to ask today. After a ROtrx's contract is evicted, how is it got back into the code cache again if it happens to be scheduled to run in the read window? Where is the code to make it wait until write window? Thanks again.

Comment (linh2931) [reply libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp:145]: I merge the PR now. Will do further work if necessary.

**Pull request merged**